### ISSUE TITLE
Add option to use windows keymap on MacOS

### ISSUE THREAD

**Issue by k--kato** [Microsoft/vscode#27385]
- VSCode Version: 1.12.2
- OS Version: macOS Sierra 

We can write `keybindings` in `package.json` (Extension Manifest File).
```json
"keybindings": [
    {
        "win": "ctrl+y",
        "mac": "cmd+backspace",
        "linux": "ctrl+y",
        "command": "editor.action.deleteLines",
        "when": "editorTextFocus && !editorReadonly"
    }
]
```
I got a request from some developers that they are using macOS but they want to use Windows keybindings. See [Add option to use windows keymap on MacOS](https://github.com/k--kato/vscode-intellij-idea-keybindings/issues/67).

So, I'd like to switch `mac` and `win` in `settings.json` or something. Can I do this?

**Comment by pwcccw** [comment]
Bump

**Comment by sumitgpt** [comment]
I too want this

**Comment by Syeberman** [comment]
Bump

**Comment by rrousselGit** [comment]
I'd like this feature as well for the same reasons.

I can contribute. But I don't know where to get started. @alexandrudima @roblourens any guidance on where to start things ?

**Comment by mredbishop** [comment]
Yes, this feature would be really useful for us too. We are predominantly a Linux studio but we do use OSX for some dev work. It would be great when using OSX we could have the Linux key bindings if possible?

**Comment by petermetz** [comment]
As an extension to this: should also work the other direction. ;-)

**Comment by festum** [comment]
Waited for 2 years. I only have to overwrite windows keymap on mac keymap setting.

**Comment by shynome** [comment]
Add option to use windows keymap on Linux

**Comment by mredbishop** [comment]
Still need to be able to switch to the defaults for another OS. This would be especially useful now as a number of us have started using WSL 2 due to the performance improvements over WSL 1 but we have trained our brains on the Linux key mappings :o)

**Comment by virhonestum** [comment]
I would also love this, as I use Linux for work and Windows on my private machine. This could also be implemented as an extension or multiple ones on the marketplace.

**Comment by CultivateCreate** [comment]
Yes. I just synced my parallels version of vscode to the settings I uploaded from my mac, and now It's completely unusable. 

### _**PLEASE implement the 'super' key as seen in Sublime.**_

**Comment by philpowers** [comment]
Lack of this functionality basically makes VSCode a non-starter for me on the Mac... My Mac sits right next to my PC, and if I have to memorize two sets of key bindings, I might as well stay in Xcode all the time (which makes me sad).

**Comment by mscottmoore** [comment]
I also would really like this functionality, as a switch back and forth between a Mac at work and Linux at home.

I looked into the code a bit, and had some luck hardcoding some OS values, but I was getting some weird behavior. We would also want a final version of this to be configurable in the preferences/settings. Unfortunately it looked like a lot of this OS stuff is happening at too low of a level to have access to the configurationService. I'm not familiar with the vscode code base, so I might be wrong about some of this, but this at least doesn't seem like a beginner-friendly issue to fix.

I did find a potential work-around though. There's a "Preferences: Open Default Keyboard Shortcuts (JSON)" command which opens a readonly editor with all of your current OS's shortcuts. You can copy the contents of that file and then use "Preferences: Open Keyboard Shortcuts" on your other computer to open your user-defined shortcuts and paste in the other OS's defaults. You could also do that on one computer and then sync your settings. I'll try that out this weekend and see how it goes.

**Comment by FredHappyface** [comment]
I've created a few extensions that should help with this. Though it would certainly be preferable to have this functionality out of the box 
```
ext install fredhappyface.windowskeybindings
```
```
ext install fredhappyface.linuxkeybindings
```
Note that some keybindings may interfere with other extensions. In that case drop an issue and I'll try and fix asap

**Comment by ucavalcante** [comment]
Same Request for linux, will be great if we can chose the main keymap theme on preferences, im current migrating from windows dev environment for linux and need to learn everything again.

**Comment by UniverseFly** [comment]
Same request for Mac. I'm a frequent Mac user but sometimes would be using Windows when I wanna stick to my muscle memory for Mac.

**Comment by duck-dev-go** [comment]
I would also like this feature

**Comment by prantlf** [comment]
You could use the [clause contexts](https://code.visualstudio.com/api/references/when-clause-contexts#available-contexts) `isLinux`, `isMac`, `isWindows` and `isWeb` as a workaround. For example:

```json
"keybindings": [
  {
    "command": "editor.action.deleteLines",
    "key": "cmd+backspace",
    "when": "isMac && editorTextFocus && !editorReadonly"
  },
  {
    "command": "editor.action.deleteLines",
    "key": "ctrl+y",
    "when": "!isMac && editorTextFocus && !editorReadonly"
  }
]
```

But your suggestion with the [property `mac` will work too](https://code.visualstudio.com/api/references/contribution-points#keybinding-example). Just use `key` instead of `win`:

```json
"keybindings": [
  {
    "command": "editor.action.deleteLines",
    "key": "ctrl+y",
    "mac": "cmd+backspace",
    "when": "isMac && editorTextFocus && !editorReadonly"
  }
]
```

**Comment by mogelbrod** [comment]
After working on vscode extensions I too discovered the ability to define platform-specific bindings and immediately went and replaced all the duplicated bindings I had in keybidings.json. I was rather surprised (and disappointed) when I discovered that they didn't work 🙁  Having support for `{ "key": "...", "mac": "...", "win": "...", "linux": "..." }` would remove the need to define and maintain otherwise identical mappings that only differ by the `key` prop as well as including `"when": "isX && ..."`. Maintaining these are error prone, and result in 2-4 times as many bindings having to be added and updated anytime one wants to change the command or when clause.

Example of verbose double mapping required today (note that multiple `isMac` checks are necessary):
```json
[
  {
    "key": "cmd+w",
    "command": "workbench.action.terminal.kill",
    "when": "terminalFocus && terminalHasBeenCreated || terminalFocus && terminalProcessSupported"
  },
  {
    "key": "ctrl+w",
    "command": "workbench.action.terminal.kill",
    "when": "!isMac && terminalFocus && terminalHasBeenCreated || !isMac && terminalFocus && terminalProcessSupported"
  },
]
```
My personal keybindings.json has ~40 of these mappings, which makes it much harder to maintain and organize them.
Compare this to the equivalent binding that the proposal would enable:
```json
[
  {
    "key": "cmd+w",
    "mac": "cmd+w",
    "command": "workbench.action.terminal.kill",
    "when": "terminalFocus && terminalHasBeenCreated || terminalFocus && terminalProcessSupported"
  },
]
```

**Comment by editicalu** [comment]
While this doesn't solve the fundamental design issue in VS Code's keybinding system, a third party extension created all keybindings from Windows for other platforms.
https://marketplace.visualstudio.com/items?itemName=smcpeak.default-keys-windows

**Comment by editicalu** [comment]
While this doesn't solve the design issue in vscode, there's an extension to (partially) solve the issue. Leaving it here as it might be useful for people having the same frustrations.
https://marketplace.visualstudio.com/items?itemName=smcpeak.default-keys-windows

**Comment by fMoro1999** [comment]
Follow

**Comment by cheyngoodman** [comment]
I'm using a Mac keyboard on both MacOS and Linux. Rebinding keybinds works for most things.

I can't find a way to rebind `Ctrl + Click` for opening links to `command + Click` to open links.

**Comment by kaiwu-astro** [comment]
2024 - 2017 = 7 years!
Microsoft guys never use multiple OS?

**Comment by fMoro1999** [comment]
@kaiwu-astro, plot twist: Microsoft devs work using MacBooks 💀

**Comment by marcospgp** [comment]
I wanted to make keybinds on Windows the same as MacOS, and got most of the way there using PowerToys:

```markdown
### Emulate MacOS keyboard shortcuts

To emulate MacOS keyboard shortcuts on Windows, use the PowerToys Keyboard Manager with the following settings:

(note that replacing left alt and left ctrl should be good enough, but to keep things short "left" is omitted from these key names below. "left" on its own means "left arrow key".)

Remap keys:

- alt -> ctrl
- ctrl -> alt

Remap shortcuts:

(note that the key remap above applies first, so alt below represents the ctrl key on the keyboard, and vice versa)

Restore switching between windows with alt + tab after key remap:

- ctrl + tab -> alt + tab

Restore switching tabs:

- alt + tab -> ctrl + tab
- alt + shift + tab -> ctrl + shift + tab

Move cursor to start/end of line (does not seem to require additional entries for selecting text with the addition of the shift key):

- ctrl + left -> Home
- ctrl + right -> End

Move cursor word by word, also selecting text if shift included:

- win + left -> ctrl + left
- win + right -> ctrl + right
- win + shift + left -> ctrl + shift + left
- win + shift + right -> ctrl + shift + right

Erase word by word:

- win + backspace -> ctrl + backspace

Drag text lines up and down:

- win + up -> alt + up
- win + down -> alt + down

Missing:

- Erase to start of line (cmd + backspace on MacOS) (not very important as one can simply select the line, then erase.)
- Some VSCode keybinds are the same between Windows and MacOS, and this remap will break them if they use ctrl or alt, so they need to be restored - such as with "toggle terminal" (JSON generated by VSCode after modifying keybind):

  ```json
    {
      "key": "alt+oem_3",
      "command": "workbench.action.terminal.toggleTerminal",
      "when": "terminal.active"
    },
    {
      "key": "ctrl+oem_3",
      "command": "-workbench.action.terminal.toggleTerminal",
      "when": "terminal.active"
    }
  ```
```
